x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/utils/ingest/PipelineRegistry.java

```diff
@@ -0,0 +1,197 @@
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.elasticsearch.core.Strings.format;
+import static org.elasticsearch.xpack.core.ClientHelper.executeAsyncWithOrigin;
+
+/**
+ * Allow to create a registry that will manage ingest pipelines setup while the service is started.
+ * Also, pipeline are versioned, allowing to keep them up to date.
+ * The work is heavily inspired by {@link org.elasticsearch.xpack.core.template.IndexTemplateRegistry}.
+ */
+public abstract class PipelineRegistry implements ClusterStateListener {
+
+    private static final Logger logger = LogManager.getLogger(PipelineRegistry.class);
+    private final ThreadPool threadPool;
+    private final Client client;
+
+    private final ConcurrentMap<String, AtomicBoolean> pipelineCreationsInProgress = new ConcurrentHashMap<>();
+
+    public PipelineRegistry(ClusterService clusterService, ThreadPool threadPool, Client client) {
+        this.threadPool = threadPool;
+        this.client = client;
+        clusterService.addListener(this);
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        ClusterState state = event.state();
+        if (state.blocks().hasGlobalBlock(GatewayService.STATE_NOT_RECOVERED_BLOCK)) {
+            // wait until recovered from disk, so the cluster state view is consistent
+            return;
+        }
+
+        DiscoveryNode masterNode = event.state().getNodes().getMasterNode();
+        if (masterNode == null || state.nodes().isLocalNodeElectedMaster() == false) {
+            // no master node elected or current node is not master
+            return;
+        }
+
+        addIngestPipelinesIfMissing(state);
+    }
+
+    protected abstract String getOrigin();
+
+    protected abstract List<PipelineTemplateConfiguration> getIngestPipelineConfigs();
+
+    private void addIngestPipelinesIfMissing(ClusterState state) {
+        for (PipelineTemplateConfiguration pipelineTemplateConfig : getIngestPipelineConfigs()) {
+            PipelineConfiguration newPipeline = pipelineTemplateConfig.load();
+            final AtomicBoolean creationCheck = pipelineCreationsInProgress.computeIfAbsent(
+                newPipeline.getId(),
+                key -> new AtomicBoolean(false)
+            );
+
+            if (creationCheck.compareAndSet(false, true)) {
+                if (ingestPipelineExists(state, newPipeline.getId())) {
+                    IngestMetadata ingestMetadata = state.metadata().custom(IngestMetadata.TYPE);
+                    PipelineConfiguration existingPipeline = ingestMetadata.getPipelines().get(newPipeline.getId());
+                    boolean newPipelineHasVersion = Objects.isNull(newPipeline.getVersion()) == false;
```

Comment:
It makes more sense right now because the code is 
```
boolean newPipelineHasVersion = Objects.nonNull(newPipeline.getVersion())
```

It removes the need of `==` and I find it easier to read.